File: session_py/src/session_py/quaternion.py
import uuid
import json
import math
from .vector import Vector


class Quaternion:
    def __init__(self, s=1.0, v=None):
        self.typ = "Quaternion"
        self.guid = str(uuid.uuid4())
        self.name = "my_quaternion"
        self.s = s
        self.v = v if v is not None else Vector(0.0, 0.0, 0.0)

    @staticmethod
    def identity():
        return Quaternion(1.0, Vector(0.0, 0.0, 0.0))

    @staticmethod
    def from_sv(s, x, y, z):
        return Quaternion(s, Vector(x, y, z))

    @staticmethod
    def from_axis_angle(axis, angle):
        axis = axis.normalize()
        half_angle = angle * 0.5
        s = math.cos(half_angle)
        v = axis * math.sin(half_angle)
        return Quaternion(s, v)

    def rotate_vector(self, v):
        qv = self.v
        uv = qv.cross(v)
        uuv = qv.cross(uv)
        return v + (uv * self.s + uuv) * 2.0

    def magnitude(self):
        return math.sqrt(
            self.s * self.s + self.v.x * self.v.x + self.v.y * self.v.y + self.v.z * self.v.z
        )

    def normalize(self):
        mag = self.magnitude()
        if mag > 1e-10:
            q = Quaternion(self.s / mag, self.v / mag)
            q.typ = self.typ
            q.guid = self.guid
            q.name = self.name
            return q
        else:
            return Quaternion.identity()

    def conjugate(self):
        q = Quaternion(self.s, self.v * -1.0)
        q.typ = self.typ
        q.guid = self.guid
        q.name = self.name
        return q

    def __mul__(self, other):
        if isinstance(other, Quaternion):
            s = self.s * other.s - self.v.dot(other.v)
            v = other.v * self.s + self.v * other.s + self.v.cross(other.v)
            return Quaternion(s, v)
        raise TypeError("Quaternion can only be multiplied with another Quaternion")

    def to_json_data(self):
        return {
            "type": self.typ,
            "guid": self.guid,
            "name": self.name,
            "s": self.s,
            "x": self.v.x,
            "y": self.v.y,
            "z": self.v.z,
        }

    @staticmethod
    def from_json_data(data):
        q = Quaternion(data["s"], Vector(data["x"], data["y"], data["z"]))
        q.typ = data.get("type", "Quaternion")
        q.guid = data["guid"]
        q.name = data["name"]
        return q

    def to_json(self, filepath):
        with open(filepath, "w") as f:
            json.dump(self.to_json_data(), f, indent=4)

    @staticmethod
    def from_json(filepath):
        with open(filepath, "r") as f:
            data = json.load(f)
            return Quaternion.from_json_data(data)

File: session_cpp/src/quaternion.cpp
#include "quaternion.h"
#include "guid.h"
#include <cmath>
#include <fstream>

namespace session_cpp {

Quaternion::Quaternion() : typ("Quaternion"), guid(::guid()), name("my_quaternion"), s(1.0f), v(0.0f, 0.0f, 0.0f) {}

Quaternion::Quaternion(float s, const Vector& v) 
    : typ("Quaternion"), guid(::guid()), name("my_quaternion"), s(s), v(v) {}

Quaternion Quaternion::identity() {
    return Quaternion(1.0f, Vector(0.0f, 0.0f, 0.0f));
}

Quaternion Quaternion::from_sv(float s, float x, float y, float z) {
    return Quaternion(s, Vector(x, y, z));
}

Quaternion Quaternion::from_axis_angle(const Vector& axis, float angle) {
    Vector axis_copy = axis;
    Vector normalized_axis = axis_copy.normalize();
    float half_angle = angle * 0.5f;
    float s = std::cos(half_angle);
    Vector v = normalized_axis * std::sin(half_angle);
    return Quaternion(s, v);
}

Vector Quaternion::rotate_vector(const Vector& vec) const {
    Vector qv = v;
    Vector vec_copy = vec;
    Vector uv = qv.cross(vec_copy);
    Vector uuv = qv.cross(uv);
    return vec_copy + (uv * s + uuv) * 2.0f;
}

float Quaternion::magnitude() const {
    return std::sqrt(s * s + v.x() * v.x() + v.y() * v.y() + v.z() * v.z());
}

Quaternion Quaternion::normalize() const {
    float mag = magnitude();
    if (mag > 1e-10f) {
        Quaternion q(s / mag, v / mag);
        q.typ = typ;
        q.guid = guid;
        q.name = name;
        return q;
    } else {
        return Quaternion::identity();
    }
}

Quaternion Quaternion::conjugate() const {
    Quaternion q(s, v * -1.0f);
    q.typ = typ;
    q.guid = guid;
    q.name = name;
    return q;
}

Quaternion Quaternion::operator*(const Quaternion& other) const {
    Vector v_copy = v;
    Vector other_v_copy = other.v;
    float new_s = s * other.s - v_copy.dot(other_v_copy);
    Vector new_v = other_v_copy * s + v_copy * other.s + v_copy.cross(other_v_copy);
    return Quaternion(new_s, new_v);
}

nlohmann::ordered_json Quaternion::to_json_data() const {
    auto clean_float = [](float val) -> double { 
        // For very small values, keep high precision
        if (std::abs(val) < 0.01f) return static_cast<double>(val);
        // For normal values, round to 2 decimal places
        return static_cast<double>(std::round(val * 100.0f) / 100.0f);
    };
    return nlohmann::ordered_json{
        {"type", typ},
        {"guid", guid},
        {"name", name},
        {"s", clean_float(s)},
        {"x", clean_float(v.x())},
        {"y", clean_float(v.y())},
        {"z", clean_float(v.z())}
    };
}

Quaternion Quaternion::from_json_data(const nlohmann::json& data) {
    Quaternion q(data["s"].get<float>(), Vector(data["x"].get<float>(), data["y"].get<float>(), data["z"].get<float>()));
    q.typ = data.value("type", "Quaternion");
    q.guid = data["guid"].get<std::string>();
    q.name = data["name"].get<std::string>();
    return q;
}

void Quaternion::to_json(const std::string& filepath) const {
    std::ofstream file(filepath);
    file << to_json_data().dump(4);
    file.close();
}

Quaternion Quaternion::from_json(const std::string& filepath) {
    std::ifstream file(filepath);
    nlohmann::json data;
    file >> data;
    file.close();
    return from_json_data(data);
}

}  // namespace session_cpp
